File: typescript-playwright/src/pages/BasePage.ts
import type { Page, Locator } from "@playwright/test";

export abstract class BasePage {
  protected readonly page: Page;

  constructor(page: Page) {
    this.page = page;
  }
  
  get footer(): Locator {
    return this.page.locator("footer");
  }

  get homeButton(): Locator {
    return this.page.locator('[data-testid="home-button"]');
  }

  protected testIdStartsWith(prefix: string): Locator {
    return this.page.locator(`[data-testid^="${prefix}"]`);
  }

  async scrollToBottom(): Promise<void> {
    await this.page.evaluate(() => {
      window.scrollTo(0, document.body.scrollHeight);
    });
  }

  async backToHome(): Promise<void> {
    await this.homeButton.click();
  }
}

File: python/src/pages/base_page.py
from playwright.async_api import Locator, Page


class BasePage:
    def __init__(self, page: Page):
        self.page = page

    @property
    def footer(self) -> Locator:
        return self.page.get_by_test_id("footer")

    @property
    def home_btn(self) -> Locator:
        return self.page.get_by_test_id("home-btn")

    def test_id_starts_with(self, prefix: str) -> Locator:
        return self.page.locator(f"[data-testid^='{prefix}']")

    async def scroll_to_bottom(self) -> None:
        await self.page.evaluate("window.scrollTo(0, document.body.scrollHeight)")
